src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.SymbolComputer.cs

```diff
@@ -9,32 +9,56 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis.Completion.Providers.ImportCompletion;
 using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.PooledObjects;
 using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Completion.Providers
 {
     internal static partial class ExtensionMethodImportCompletionHelper
     {
-        private class ExtensionMethodSymbolComputer
+        private class SymbolComputer
         {
             private readonly int _position;
             private readonly Document _originatingDocument;
             private readonly SemanticModel _originatingSemanticModel;
             private readonly ITypeSymbol _receiverTypeSymbol;
             private readonly ImmutableArray<string> _receiverTypeNames;
             private readonly ISet<string> _namespaceInScope;
-            private readonly IImportCompletionCacheService<CacheEntry, object> _cacheService;
+            private readonly IImportCompletionCacheService<ExtensionMethodImportCompletionCacheEntry, object> _cacheService;
 
             // This dictionary is used as cache among all projects and PE references. 
             // The key is the receiver type as in the extension method declaration (symbol retrived from originating compilation).
             // The value indicates if we can reduce an extension method with this receiver type given receiver type.
             private readonly ConcurrentDictionary<ITypeSymbol, bool> _checkedReceiverTypes;
 
-            public ExtensionMethodSymbolComputer(
+            private static readonly object _gate = new();
+            private static AsyncBatchingWorkQueue<Project>? _workQueue;
```

Comment:
this is a bit wonky.  see the comment in https://github.com/dotnet/roslyn/pull/59760/files#r830328316 for a likely cleaner way of doing this.  you won't need a lock, and you won't need to do weird things to get teh AsyncOpListener.